pool/native.go

```diff
@@ -0,0 +1,159 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// NativePoolOptions specify options for NativePool.
+type NativePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      uint
+	Type      reflect.Type
+}
+
+// NativePool represents an object pool which is opaque to the Go GC.
+type NativePool interface {
+	Get() interface{}
+	GetOr(OverflowFn) interface{}
+	Put(interface{})
+
+	// Owns determines if the object belongs to the pool.
+	Owns(interface{}) bool
+
+	// Size returns the used and the total capacity of the pool.
+	Size() (uint64, uint64)
+}
+
+// OverflowFn produces non-pooled objects.
+type OverflowFn func() interface{}
+
+// NewNativePool constructs a new NativePool.
+func NewNativePool(opts NativePoolOptions) NativePool {
+	p := &nativePool{
+		free: make(chan uint64, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: uint64(hsz + opts.Type.Size())}
+
+	p.size = uint64(p.opts.Size) * p.step
+
+	if p.size == 0 {
+		panic("native-pool: pool size is zero")
+	}
+
+	if p.step%uint64(p.opts.Type.Align()) != 0 {
+		panic("native-pool: invalid alignment")
+	}
+
+	if p.opts.Construct == nil {
+		p.opts.Construct = func(interface{}) {}
+	}
+
+	p.init()
+
+	return p
+}
+
+type hdr struct {
+	// Offset from the beginning of the arena to the object.
+	idx uint64
+}
+
+// Header size with padding, as determined by the compiler.
+const hsz = unsafe.Sizeof(*(*hdr)(nil))
+
+type nativePool struct {
+	pool       []byte
+	free       chan uint64
+	opts       NativePoolOptions
+	step, size uint64
+}
+
+func (p *nativePool) init() {
+	// Heap is a slice of bytes large enough to fit opts.Size objects
+	// of type struct { hdr; T }.
+	p.pool = mmap(int(p.size))
+
+	for i := uint64(0); i < p.size; i += p.step {
+		hdr := (*hdr)(unsafe.Pointer(&p.pool[i]))
+		hdr.idx = i + uint64(hsz)
+		ptr := unsafe.Pointer(&p.pool[hdr.idx])
+
+		p.opts.Construct(reflect.NewAt(p.opts.Type, ptr).Interface())
+
+		// Alternatively, we can use a channel of interfaces, but it
+		// would keep opts.Size objects rooted in the channel.
+		// We aim to avoid allocating unnecessary GC-visible objects.
+		p.free <- hdr.idx
+	}
+}
+
+func (p *nativePool) Size() (uint64, uint64) {
+	free := uint64(cap(p.free)) * p.step
+	used := p.size - uint64(len(p.free))*p.step
+
+	return used, free
+}
+
+// Get provides an object from the pool.
+func (p *nativePool) Get() interface{} {
+	return reflect.NewAt(
+		p.opts.Type, unsafe.Pointer(&p.pool[<-p.free])).Interface()
```

Comment:
Yes, it will. Yes, we use it in NativeHeap to avoid the `select` overhead when we know that there are free objects in the pool.